src/aind_codeocean_utils/codeocean_job.py

```diff
@@ -0,0 +1,495 @@
+            )
+            logger.info(
+                f"Updating permissions {update_res_perm_response.status_code}"
+            )
+        return reg_result_response
+
+    def process(self) -> dict:
+        """
+        Method to process data with a Code Ocean capsule/pipeline.
+        """
+        responses = dict()
+
+        # 1. register data assets (optional)
+        if self.job_config.register_config is not None:
+            logger.info("Registering data asset")
+            register_data_asset_response = (
+                self._register_data_and_update_permissions(
+                    self.job_config.register_config
+                )
+            )
+            self.job_config.run_capsule_config.data_assets.append(
+                dict(
+                    id=register_data_asset_response.json()["id"],
+                    mount=self.job_config.register_config.mount,
+                )
+            )
+            data_asset_tags = self.job_config.register_config.tags
+            data_asset_custom_metadata = deepcopy(
+                self.job_config.register_config.custom_metadata
+            )
+            responses["register"] = register_data_asset_response
+        else:
+            if len(self.job_config.run_capsule_config.data_assets) == 1:
+                data_asset = self.job_config.run_capsule_config.data_assets[0]
+                data_asset_json = self.co_client.get_data_asset(
+                    data_asset["id"]
+                ).json()
+                data_asset_tags = data_asset_json["tags"]
+                data_asset_custom_metadata = data_asset_json["custom_metadata"]
+            else:
+                # tags and custom_metadata are not propagated if multiple data
+                # assets are provided
+                data_asset_tags = []
+                data_asset_custom_metadata = {}
+
+        # 2. run capsule
+        logger.info("Running capsule")
+        run_capsule_response = self._run_capsule(
+            self.job_config.run_capsule_config
+        )
+        responses["run"] = run_capsule_response
+
+        # 3. capture results (optional)
+        if self.job_config.capture_result_config is not None:
+            logger.info("Capturing results")
+            result_tags = [
+                DataLevel.DERIVED.value if x == DataLevel.RAW.value else x
+                for x in data_asset_tags
+            ]
+            self.job_config.capture_result_config.tags.extend(result_tags)
+            results_metadata = deepcopy(data_asset_custom_metadata)
+            results_metadata["data level"] = "derived data"
+            self.job_config.capture_result_config.custom_metadata.update(
+                results_metadata
+            )
+            capture_result_response = self._capture_result(
+                computation_id=run_capsule_response.json()["id"],
+                input_data_asset_name=(
+                    self.job_config.register_config.asset_name
```

Comment:
@alejoe91 I get an error in the event that the Optional register_config is None. What's this input_data_asset_name supposed to be?

Reply:
nice catch, pushing a fix!

Reply:
Fixed. Basically you have 2 cases when no registration is done:
1. there is a single data asset attached to the run capsule --> retrieve its name as `input_data_asset_name`
2. there are multiple data assets attached --> `input_data_asset_name` needs to be specified

Note that 1. is not tested because it requires a call to the CO API to get the asset name associated to the ID. 
Can we use a mock here too?

Reply:
Yep. I mock all the responses from code ocean in the unit tests